Add new FIDO2/Passkey deep dive (#290)
bitwarden/contributing-docs: docs/architecture/deep-dives/passkeys/index.md

# FIDO2 and Passkeys

:::info

This page provides an overview of the FIDO2 standard and its architecture. It is intended to provide
a generic high-level understanding of FIDO2. For more information on how Bitwarden implements FIDO2,
see [Implementations](./implementations).

> :::

Author: Yes but @withinfocus [commented](https://github.com/bitwarden/contributing-docs/pull/290#discussion_r1525335659) that we should remove these types of sections, while I still feel that this one provides some value, so I thought that an admonition would strike good middle ground.

I'm all for changing this to regular text

> :::

Reviewer: I think removing "This page provides an overview of the FIDO2 standard and its architecture. It is intended to provide
a generic high-level understanding of FIDO2." would be sufficient, and then this is a nice info block to call out the other page.

## Architecture

FIDO2 can be broken down into two main components: WebAuthn and CTAP2. WebAuthn is a web standard
that allows for the creation and use of passkeys through a well-defined JavaScript interface, and
CTAP2 is a protocol for communicating with external authenticators (also know as roaming
authenticators), such as hardware security keys (e.g. YubiKeys).

The interaction between these two components can vary depending on the use case:

- When using platform credentials, the interaction is always between the browser and the platform
  authenticator (i.e. the operating system) using native APIs.
- When using roaming authenticators, the interaction can be either:
  - Mediated by the platform, or
  - Directly between the browser and the authenticator using e.g. USB/HID or BLE protocols.

### Diagram

The following diagram shows a generic overview of the FIDO2 architecture. The diagram is a
high-level overview and does not include any specific details about how Bitwarden implements FIDO2.

:::info

The `JavaScript Application` is also part of the `Relying Party`, even if the diagram does not
reflect this.

> :::

Author: I did my best but I could not find a way to get a box around a nested element from another box.

Could you maybe draw me a quick doodle? xD

> :::

Reviewer 2: Something like this, excuse the bad mockup
![image](https://github.com/bitwarden/contributing-docs/assets/137855/048bdbc6-f331-44ab-8351-99c123f89c60)

# FIDO2 and Passkeys

This is not a comprehensive guide to FIDO2, but a summary of the standard's crucial aspects for

> working within the Bitwarden context.

Reviewer: 🎨 Instead of this being a mostly-empty page, place all the content on the "overview" page here. I think this flows better and seems to be what other deep dives do, plus it's less time spent navigating.

A couple of the architecture pages have a sub-page with an overview, but I think that's actually an area for improvement.

# FIDO2 and Passkeys

This is not a comprehensive guide to FIDO2, but a summary of the standard's crucial aspects for
working within the Bitwarden context.

> It is not intended to be a comprehensive guide to FIDO2, but rather a summary of the most important

Reviewer 3: ⛏️ This appears to be largely duplicative between the two paragraphs.

Author: I think I forgot to remove the second one. Fixed!

> It is not intended to be a comprehensive guide to FIDO2, but rather a summary of the most important
> aspects of the standard, to be able to work with it in the context of Bitwarden.

Reviewer 4: Nit: clarity

```suggestion
This is not a comprehensive guide to FIDO2, but a summary of the standard's crucial aspects for working within the Bitwarden context.
```
